sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/SubexpressionEliminationSuite.scala

```diff
@@ -309,6 +309,22 @@ class SubexpressionEliminationSuite extends SparkFunSuite with ExpressionEvalHel
       CodeGenerator.compile(code)
     }
   }
+
+  test("SPARK-35410: SubExpr elimination should not include redundant child exprs " +
+    "for conditional expressions") {
```

Comment:
is this only a problem for conditional expression?

Reply:
So far the only one I can think about.

Reply:
Found a non-conditional example that still is an issue even with this update (a bit contrived, but I'm sure there's a real use case)

```
val myUdf = udf(() => {
  println("In UDF")
  1
}).withName("myUdf")
spark.range(1).withColumn("a", myUdf()).select(($"a" + $"a") / ($"a" + $"a")).show()
```
This generates subexpressions myUdf() and (myUdf() + myUdf()), even though only the second one is used.

Reply:
Thanks @Kimahriman. I see. Let me also look at it. As it is non-conditional case, but looks like the similar case. Let me see if it can be solved similarly.

Reply:
Oh, I figured out. This might be an issue since we have sub-expr elimination. We also need to remove redundant children exprs for non-conditional cases.

Reply:
But the fix might be different. I will work on it locally and submit another fix for it.

Reply:
Any more thoughts on this? Was the subexpr sorting supposed to address this?

Reply:
It might need another fix. I'm working on it and will submit it after these PRs merged.